### PR TITLE
Support kerberos and delegation-token

### PR DESCRIPTION
**Pull request by zzzzming95**
### :pencil: Description

The authentication of our environment is all based on kerberos. I can see the kerberos patch in : https://github.com/ExpediaGroup/waggle-dance/commit/a4ea6950eb755ca279e405b3730a6f9a0a28f1dc

But the earlier patch for kerberos has some flaws, such it not support delegation-token auth , using the duplicate `HadoopThriftAuthBridge.Server`. We fix it and now it work well in our env.

### :link: Related Issues
https://github.com/ExpediaGroup/waggle-dance/issues/264

### CODE REVIEW

**Comment by patduin** [comment]
> @abhimanyugupta07 @javsanbel2 cc

Thanks for the contribution! We'll try to have some eyes on this asap. 
Just a quick FYI we can run our tests but we don't have a Kerberos environment ourselves so those specific changes we'll have to rely on your tests.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 1]
```diff
@@ -87,8 +99,26 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
...
           return Lists.newArrayList(user);
         }
+      case "get_delegation_token":
+        try {
+          base.open(cachedUgi);
+          Object token = doRealCall(method, args, attempt);
+          if (useSasl) {
+            this.delegationToken = (String) token;
+            base.close();
+            String newTokenSignature = base.getHiveConfValue(HiveConf.ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
```
Comment (abhimanyugupta07) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:109]: why do you have to set the default value "WAGGLEDANCETOKEN" here ?

are you not able to set it in the hive-site.xml and use it here ?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 2]
```diff
@@ -132,6 +162,33 @@ private void reconnectIfDisconnected() {
...
+      UserGroupInformation currUser = UserGroupInformation.getCurrentUser();
+      if (delegationToken == null && currUser != UserGroupInformation.getLoginUser() && useSasl) {
+        String currShortName = currUser.getShortUserName();
+        Method getTokenMethod =
+                ThriftHiveMetastore.Iface.class.getMethod("get_delegation_token",
+                        String.class, String.class);
+        Object[] args = new Object[2];
+        args[0] = currShortName;
+        args[1] = currShortName;
+        String token = (String) doRealCall(getTokenMethod, args, 0);
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:177]: This next few lines are duplicated above maybe we can refactor this in a method

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 3]
```diff
@@ -132,6 +162,33 @@ private void reconnectIfDisconnected() {
...
+        base.close();
+        String newTokenSignature = base.getHiveConfValue(HiveConf.ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
+                tokenSignature);
+        base.setHiveConfValue(HiveConf.ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
+                newTokenSignature);
+        Utils.setTokenStr(currUser, token, newTokenSignature);
+        this.delegationToken = token;
+      }
+      if (!base.isOpen()) {
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:186]: move this in inside the if above so we don't change existing logic.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java [hunk 4]
```diff
@@ -92,6 +92,12 @@ public PrefixBasedDatabaseMappingService(
 
   private void add(AbstractMetaStore metaStore) {
     MetaStoreMapping metaStoreMapping = metaStoreMappingFactory.newInstance(metaStore);
+    // connect metastore first
+    if (!metaStoreMapping.isAvailable()) {
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:96]: why was this added?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.hive.service.auth.HiveAuthFactory;
+import org.apache.hive.service.auth.PlainSaslHelper;
+import org.apache.hive.service.auth.SaslQOP;
+import org.apache.thrift.transport.TSaslServerTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.apache.thrift.transport.TTransportFactory;
+
+public class SaslHelper {
+  public static HadoopThriftAuthBridge.Server createSaslServer(HiveConf conf) throws TTransportException {
+    HadoopThriftAuthBridge.Server saslServer = saslServer = null;
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java:42]:
```
HadoopThriftAuthBridge.Server saslServer = null;
```

**Review by patduin — CHANGES_REQUESTED** [review]
Hi this looks good, just some small comments please have a look. Once merged we can get roll a release.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java (hunk 4 above)
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:96]: Hmm not sure I follow, this method just adds a Metastore to the list of configured metastores and you're just logging a warning. 
You've already changed the code to do the delegate connection here:
[DefaultMetaStoreClientFactory.java](https://github.com/ExpediaGroup/waggle-dance/pull/266/files#diff-ee675b8d870d29e0855316cdace45011de4a0a97a308fe7cefbabb4937baa63a)
Which is the correct place to add the logic. 

Adding this here just forces a client.isOpen/getStatus call earlier in the process. Might not be needed...

Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:96]: The call `StaticDatabaseMappingService` is done because we have to get the database e.g. make a real connection.
Here I think we can defer till we get a use call (Which will do the getDelegate then) but it might not have a lot of impact. We can leave it an see.

**Review by patduin — APPROVED** [review]
@abhimanyugupta07 have a look please if ok, we can merge this.

File: HowToKerberize.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
+![Bee waggle-dancing on a hive.](logo.png "Federating Hive Meta Stores.")
+
+# Additional instruction to use Waggle Dance in kerberized environment
```
Comment (massdosage) [inline HowToKerberize.md:3]:
```suggestion
# Additional instructions to use Waggle Dance in a Kerberized environment
```

File: HowToKerberize.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+
+### Process
+
+In a kerberos environment, after a client requests waggle-dance, waggle-dance requests the proxy user's token from the metastore and then uses that token to communicate with the metastore.
```
Comment (massdosage) [inline HowToKerberize.md:8]:
```suggestion
In a Kerberos environment a client make a request to Waggle Dance which in turn requests the proxy user's token from the metastore and then uses this token to communicate with the metastore.
```

File: HowToKerberize.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+This is necessary in some scenarios that require permission authentication. Such as the `create_table` API that requires the proxy user to create hdfs directories.
```
Comment (massdosage) [inline HowToKerberize.md:10]:
```suggestion
This is necessary in certain scenarios that need authentication - for example the `create_table` API that requires the proxy user to create HDFS directories.
```

File: HowToKerberize.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+![image](https://user-images.githubusercontent.com/13965087/229339323-260f3c17-46c0-4471-81d2-cdbcfa0fe3ce.png)
```
Comment (massdosage) [inline HowToKerberize.md:12]: We should move a cleaned up version of this diagram to this repository instead of an external link.

File: HowToKerberize.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+In addition, because kerberos authentication requires a delegation-token to proxy as other users. The proxy user of the session is shared globally, which means we need to make all Hive Metastores share a set of delegation-token storage so that a single delegation-token can be authenticated by multiple Metastores.
```
Comment (massdosage) [inline HowToKerberize.md:14]:
```suggestion
In addition, because Kerberos authentication requires a delegation-token to proxy as other users, the proxy user of the session is shared globally. This means we need to make all Hive Metastores share a set of delegation-token storage so that a single delegation-token can be authenticated by multiple Metastores.
```

File: HowToKerberize.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+**One solution is to use zookeeper to store tokens for all Hive Metastores, which is essential.**
```
Comment (massdosage) [inline HowToKerberize.md:16]:
```suggestion
**One solution is to use Zookeeper to store tokens for all Hive Metastores**
```

File: HowToKerberize.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+### Prerequisites
+
+* Kerberized claster:
```
Comment (massdosage) [inline HowToKerberize.md:20]:
```suggestion
* Kerberized cluster:
```

File: HowToKerberize.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  active KDC,
+  some required properties in configuration files of hadoop services
```
Comment (massdosage) [inline HowToKerberize.md:22]:
```suggestion
  some required properties in configuration files of Hadoop services
```

File: HowToKerberize.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+* User account with privileges in ipa
```
Comment (massdosage) [inline HowToKerberize.md:23]: What is `ipa`?

File: HowToKerberize.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+* Zookeeper to store delegation-token (Recommend)
```
Comment (massdosage) [inline HowToKerberize.md:24]:
```suggestion
* Zookeeper to store delegation-token (Recommended)
```

File: HowToKerberize.md [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+
+### Configuration
+
+Waggle Dance does not read hadoop's *core-site.xml* so a general property providing kerberos auth should be added to
+the Hive configuration file *hive-site.xml*:
```
Comment (massdosage) [inline HowToKerberize.md:29]:
```suggestion
Waggle Dance does not read Hadoop's `core-site.xml` so a general property providing Kerberos auth should be added to
the Hive configuration file `hive-site.xml`:
```

File: HowToKerberize.md [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+```
+<property>
+  <name>hadoop.security.authentication</name>
+  <value>KERBEROS</value>
+</property>
+```
+
+
+Besides Waggle Dance needs a keytab file to communicate with the Metastore so following properties should be present:
```
Comment (massdosage) [inline HowToKerberize.md:39]:
```suggestion
Waggle Dance also needs a keytab file to communicate with the Metastore so the following properties should be present:
```

File: HowToKerberize.md [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+```
+<property>
+  <name>hive.metastore.sasl.enabled</name>
+  <value>true</value>
+</property>
+<property>
+  <name>hive.metastore.kerberos.principal</name>
+  <value>hive/_HOST@DEV.DF.SBRF.RU</value>
```
Comment (massdosage) [inline HowToKerberize.md:47]: Can we make this more generic?

File: HowToKerberize.md [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+</property>
+<property>
+  <name>hive.metastore.kerberos.keytab.file</name>
+  <value>/etc/hive.keytab</value>
+</property>
+```
+
+In addition, all metastores need to use the zk shared token:
```
Comment (massdosage) [inline HowToKerberize.md:55]:
```suggestion
In addition, all metastores need to use the Zookeeper shared token:
```

File: HowToKerberize.md [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    <name>hive.cluster.delegation.token.store.zookeeper.connectString</name>
+    <value>zk1:2181,zk2:2181,zk3:2181</value>
+  </property>
+  <property>
+    <name>hive.cluster.delegation.token.store.zookeeper.znode</name>
+    <value>/hive/token</value>
+  </property>
+```
+
+If you are intending to use a beeline client, following properties may be valuable:
```
Comment (massdosage) [inline HowToKerberize.md:71]:
```suggestion
If you are intending to use a Beeline client, the following properties may be valuable:
```

File: HowToKerberize.md [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+  <name>hive.server2.transport.mode</name>
+  <value>http</value>
+</property>
+<property>
+  <name>hive.server2.authentication</name>
+  <value>KERBEROS</value>
+</property>
+<property>
+  <name>hive.server2.authentication.kerberos.principal</name>
+  <value>hive/_HOST@DEV.DF.SBRF.RU</value>
```
Comment (massdosage) [inline HowToKerberize.md:83]: Same as above

File: HowToKerberize.md [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+<property>
+  <name>hive.server2.enable.doAs</name>
+  <value>false</value>
+</property>
+```
+
+
+### Running
+
+The Waggle Dance should be starting as privileged user with a fresh keytab.
```
Comment (massdosage) [inline HowToKerberize.md:98]:
```suggestion
Waggle Dance should be started by a privileged user with a fresh keytab.
```

File: HowToKerberize.md [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+
+If Waggle Dance throws a GSS exception, you have problem with the keytab file.
+Try to perform `kdestroy` and `kinit` operations and check for a keytab file ownership.
```
Comment (massdosage) [inline HowToKerberize.md:101]:
```suggestion
Try to perform `kdestroy` and `kinit` operations and check the keytab file ownership flags.
```

File: HowToKerberize.md [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+
+If the Metastore throws an exception with code -127, Waggle Dance uses wrong authentication policy.
```
Comment (massdosage) [inline HowToKerberize.md:103]:
```suggestion
If the Metastore throws an exception with code -127, Waggle Dance is probably using the wrong authentication policy.
```

File: HowToKerberize.md [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+Check hive-conf.xml for a correct configuration and make sure that HIVE_HOME and HIVE_CONF_DIR are defined.
```
Comment (massdosage) [inline HowToKerberize.md:104]:
```suggestion
Check the values in `hive-conf.xml` and make sure that HIVE_HOME and HIVE_CONF_DIR are defined.
```

File: waggle-dance-core/pom.xml [hunk 26]
```diff
@@ -350,6 +350,21 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
```
Comment (massdosage) [inline waggle-dance-core/pom.xml:353]: This section is for `test` dependencies. Is this only needed for tests? If so `<scope>test</scope>` should be added. If not it should be moved to the `hive` section between lines 118 and 163.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.hadoop.hive.thrift.HadoopThriftAuthBridge;
+import org.apache.hadoop.hive.thrift.HiveDelegationTokenManager;
+import org.apache.hive.service.auth.HiveAuthFactory;
+import org.apache.hive.service.auth.PlainSaslHelper;
+import org.apache.hive.service.auth.SaslQOP;
+import org.apache.thrift.transport.TSaslServerTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.apache.thrift.transport.TTransportFactory;
+
+public class SaslHelper {
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java:40]: Since this is a "util" class it should be made final and have a private no-args constructor added to enforce only static usage.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  public static Map<String, String> getSaslProperties(HiveConf hiveConf) {
+    Map<String, String> saslProps = new HashMap<String, String>();
+    SaslQOP saslQOP = SaslQOP.fromString(hiveConf.getVar(HiveConf.ConfVars.HIVE_SERVER2_THRIFT_SASL_QOP));
+    saslProps.put(Sasl.QOP, saslQOP.toString());
+    saslProps.put(Sasl.SERVER_AUTH, "true");
+    return saslProps;
+  }
+}
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/util/SaslHelper.java:117]: Please add a closing line break.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 29]
```diff
@@ -132,6 +158,36 @@ private void reconnectIfDisconnected() {
...
+        this.delegationToken = token;
+        if (!base.isOpen()) {
+          base.open(cachedUgi);
+        }
+      }
+      return delegationToken;
+    }
+
+    private void setTokenStr2Ugi(UserGroupInformation currUser, String token) throws IOException {
+      String newTokenSignature = base.getHiveConfValue(HiveConf.ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:185]: I'd suggest replacing the `getHiveConfValue` and `setHiveConfValue` methods could with a new method, something like `generateNewTokenSignature` which does all of this?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 30]
```diff
@@ -52,6 +60,10 @@ private ReconnectingMetastoreClientInvocationHandler(
       this.name = name;
       this.maxRetries = maxRetries;
       this.base = base;
+      if (base != null)
+        this.useSasl = base.getHiveConfBool(HiveConf.ConfVars.METASTORE_USE_THRIFT_SASL);
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:64]: We could rename this method to `isSaslEnabled`, remove the argument and then we won't need to expose the fact that the conf is being used etc.

**Review by massdosage — CHANGES_REQUESTED** [review]

**Comment by patduin** [comment]
> @abhimanyugupta07 have a look please if ok, we can merge this.

or @massdosage  :) thanks for reviewing

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java (hunk 1 above)
Comment (abhimanyugupta07) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java]: can you elaborate a bit more on why do you have to mask a default value for this property ?

Comment (abhimanyugupta07) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java]: If you put a null/empty check and only set it when it is actually set in the hive-site.xml, will it not be sufficient ?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java (hunk 4 above)
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:96]: Hi been testing with this call and please remove this. It will impact normal calls.
Considers federating to 10+ metastores we'll be doing this isAvailable check for all of them and if any of those are slow to respond we've added unnecessary latency. These calls should happen lazily only if needed. 
Which I think will work ok because of your changes in the `DefaultMetaStoreClientFactory`.

**Review by patduin — CHANGES_REQUESTED** [review]
Did some checks please have a look :)

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java (hunk 4 above)
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:96]: This part of the code should not have any knowledge of SASL or non SASL connection it should treat it all the same.

If this call is needed for Kerberos then something might need to be added in: `TProcessorFactorySaslDecorator`
Specifically I am thinking that perhaps you can inject: `NotifyingFederationService notifyingFederationService` and ` MetaStoreMappingFactory metaStoreMappingFactory` in the `MetaStoreProxyServer`

Then add it to call here: https://github.com/ExpediaGroup/waggle-dance/pull/266/files#diff-b9afe479963c18cd3af539fa88f9b7b728510cf1ad2b24a2ee26658a423770c1R210

In the `TProcessorFactorySaslDecorator.getProcess()`
Do a loop over notifyingFederationService.getAll() for each create a mapping and check availability. That way the extra work is only done for kerberos and won't affect anything else.

File: HowToKerberize.md (hunk 9 above)
Comment (patduin) [reply HowToKerberize.md:12]: I think what @massdosage means is to upload the image as part of the PR and reference it just like we do with: https://github.com/ExpediaGroup/waggle-dance/blob/main/system-diagram.png

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java [hunk 31]
```diff
@@ -92,10 +93,16 @@ public PrefixBasedDatabaseMappingService(
 
   private void add(AbstractMetaStore metaStore) {
     MetaStoreMapping metaStoreMapping = metaStoreMappingFactory.newInstance(metaStore);
+
     DatabaseMapping databaseMapping = createDatabaseMapping(metaStoreMapping);
 
     if (metaStore.getFederationType() == PRIMARY) {
       primaryDatabaseMapping = databaseMapping;
+      if (!metaStoreMapping.isAvailable()) {
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:101]: ok I still think this is brittle because the code relies on a side-effect of `isAvailable` but ok if this is a simple solution for kerberizing.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java [hunk 32]
```diff
@@ -430,6 +437,7 @@ public void close() throws IOException {
       for (MetaStoreMapping metaStoreMapping : mappingsByPrefix.values()) {
         metaStoreMapping.close();
       }
+      TokenWrappingHMSHandler.removeToken();
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java:440]: this line should go: it's untestable and there is lifecycle where this is calls via the `TTransportMonitor` which call closes on the TokenWrappingHMSHandler which should remove the token already.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/StaticDatabaseMappingService.java [hunk 33]
```diff
@@ -408,6 +409,7 @@ public void close() throws IOException {
       for (MetaStoreMapping metaStoreMapping : mappingsByMetaStoreName.values()) {
         metaStoreMapping.close();
       }
+      TokenWrappingHMSHandler.removeToken();
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/StaticDatabaseMappingService.java:412]: please remove this calls should no know anything about tokens

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TSetIpAddressProcessorFactory.java [hunk 34]
```diff
@@ -60,8 +60,11 @@ public TProcessor getProcessor(TTransport transport) {
       CloseableIHMSHandler baseHandler = federatedHMSHandlerFactory.create();
       IHMSHandler handler = newRetryingHMSHandler(ExceptionWrappingHMSHandler.newProxyInstance(baseHandler), hiveConf,
           false);
+      boolean useSASL = hiveConf.getBoolVar(HiveConf.ConfVars.METASTORE_USE_THRIFT_SASL);
+      IHMSHandler tokenHandler = newRetryingHMSHandler(TokenWrappingHMSHandler.newProxyInstance(handler, useSASL), hiveConf,
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TSetIpAddressProcessorFactory.java:64]: wrap this line in: 
```
if (useSasl) {
IHMSHandler tokenHandler = newRetryingHMSHandler(TokenWrappingHMSHandler.newProxyInstance(handler), hiveConf, false);
transportMonitor.monitor(transport, tokenHandler);
return new TSetIpAddressProcessor<>(tokenHandler);
} else {
 transportMonitor.monitor(transport, baseHandler);
 return new TSetIpAddressProcessor<>(handler);
}

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 35]
```diff
@@ -87,7 +100,22 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
...
+            base.close();
+            setTokenStr2Ugi(UserGroupInformation.getCurrentUser(), (String) token);
+            base.open(cachedUgi);
+          }
+          return token;
+        } catch (IOException | TException e) {
+          throw new MetastoreUnavailableException("Couldn't setup delegation token in the ugi: ", e);
+        }
       default:
+        genToken();
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:118]: shouldn't this only be done when useSasl=true?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TokenWrappingHMSHandler.java [hunk 36]
```diff
@@ -0,0 +1,93 @@
...
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.apache.hadoop.hive.metastore.IHMSHandler;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TokenWrappingHMSHandler implements InvocationHandler {
```
[the remark this answers is not in the document]
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TokenWrappingHMSHandler.java:27]: Would be nice if it had a unit test :)

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactoryTest.java [hunk 37]
```diff
@@ -107,7 +108,7 @@ public void defaultMethodCallThrowsTransportExceptionRetries() throws TException
 
     String result = iface.getName();
     assertThat(result, is("ourName"));
-    verify(base).open(null);
+    verify(base, Mockito.times(2)).open(null);
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactoryTest.java:111]: I think is back to 1 call???

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 38]
```diff
@@ -132,6 +160,22 @@ private void reconnectIfDisconnected() {
       }
     }
 
+    private void genToken() throws Throwable {
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:163]: I quite like the approach of the TokenWrappingHMSHandler being separate for this code the client I suggest we do the same and create a `SaslMetastoreClientHander implements  InvocationHandler`  
This class newInstance methos can check if useSasl=true and wrap the `ReconnectingMetastoreClientInvocationHandler` inside the Sasl handler. That way we don't affect the non sasl code paths at all.

Let me know what you think.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 39]
```diff
@@ -155,9 +226,17 @@ CloseableThriftHiveMetastoreIface newInstance(
...
       ThriftMetastoreClientManager base) {
     ReconnectingMetastoreClientInvocationHandler reconnectingHandler = new ReconnectingMetastoreClientInvocationHandler(
-        name, reconnectionRetries, base);
-    return (CloseableThriftHiveMetastoreIface) Proxy
-        .newProxyInstance(getClass().getClassLoader(), INTERFACES, reconnectingHandler);
+            name, reconnectionRetries, base);
+    if (base.isSaslEnabled()) {
+      CloseableThriftHiveMetastoreIface ifaceReconnectingHandler = (CloseableThriftHiveMetastoreIface) Proxy
+              .newProxyInstance(getClass().getClassLoader(), INTERFACES, reconnectingHandler);
+      // Warpping the SaslMetastoreClientHander to handle delegation token if using sasl
```
[the remark this answers is not in the document]
Comment (patduin) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:233]: Warpping=wrapping

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 40]
```diff
@@ -155,9 +226,17 @@ CloseableThriftHiveMetastoreIface newInstance(
       int reconnectionRetries,
       ThriftMetastoreClientManager base) {
     ReconnectingMetastoreClientInvocationHandler reconnectingHandler = new ReconnectingMetastoreClientInvocationHandler(
-        name, reconnectionRetries, base);
-    return (CloseableThriftHiveMetastoreIface) Proxy
-        .newProxyInstance(getClass().getClassLoader(), INTERFACES, reconnectingHandler);
+            name, reconnectionRetries, base);
+    if (base.isSaslEnabled()) {
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java:230]: we have a test `DefaultMetastoreClientFactory.closeNullBase` which test that base == null, this test I think can be deleted we don't check for null anywhere except in the close method.

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactoryTest.java [hunk 41]
```diff
@@ -122,7 +123,7 @@ public void set_ugi_before_call() throws Exception {
     String name = iface.getName();
 
     assertThat(name, is("ourName"));
-    verify(base).open(TEST_ARGS);
+    verify(base, Mockito.times(2)).open(TEST_ARGS);
```
Comment (patduin) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactoryTest.java:126]: this is called once

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TokenWrappingHMSHandler.java [hunk 42]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.waggledance.server;
+
+import java.lang.reflect.*;
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/TokenWrappingHMSHandler.java:18]: organize imports please

**Review by patduin — APPROVED** [review]

**Comment by patduin** [comment]
Thank you @zzzzming95 This looks good!
 I think @massdosage still has some open comments, please have a look.
@massdosage @abhimanyugupta07 if you both are happy please approve?

**Comment by massdosage** [comment]
LGTM, thanks for the contribution @zzzzming95

**Review by abhimanyugupta07 — APPROVED** [review]
